### PR TITLE
AggregateExpression Implementation

### PR DESCRIPTION
**Pull request by czpmango**
This PR implements the AggFun class refactoring as an AggregateExpression class (AggregateExpression will inherit from Expression), based on the following considerations:

- To facilitate the implementation of nested calls between AggregateExpression and other expressions, AggregateExpression will be passed in as a parameter, such as abs(AVG ($-.age)+1)

- Unify AggregateExpression processing behavior with other expressions in the Validator /opt phase. For example, a yieldValidator does not require a separate AGG-related data structure to be set. Polymorphic Expression can also be used in general treatment, such as DeduceTypeVisitor,RewriteExprVisitor and so on

*This PR is dependent on [https://github.com/vesoft-inc/nebula-graph/pull/503](https://github.com/vesoft-inc/nebula-graph/pull/503)*

### CODE REVIEW

File: src/common/expression/AggregateExpression.cpp [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+    // Read arg_
+    arg_ = decoder.readExpression();
+}
+
+const Value& AggregateExpression::eval(ExpressionContext& ctx) {
+    auto iter = nameIdMap_.find(name_->c_str());
+    if (iter == AggregateExpression::nameIdMap_.end()) {
+        return Value::kNullBadData;
+    }
```
Comment (CPWstatic) [inline src/common/expression/AggregateExpression.cpp:65]: Check outside the eval.

File: src/common/expression/AggregateExpression.cpp [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    auto uniques = result_->uniques();
+    if (distinct_) {
+        if (uniques->contains(val)) {
+            return result_->res();
+        }
+        uniques->values.emplace(val);
+    }
+
+    auto apply = aggFunMap_[iter->second];
```
Comment (CPWstatic) [inline src/common/expression/AggregateExpression.cpp:80]: ditto.

File: src/common/expression/AggregateExpression.cpp [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+const Value& AggregateExpression::eval(ExpressionContext& ctx) {
+    auto iter = nameIdMap_.find(name_->c_str());
+    if (iter == AggregateExpression::nameIdMap_.end()) {
+        return Value::kNullBadData;
+    }
+
+    auto val = arg_->eval(ctx);
+    if (!result_) {
```
Comment (CPWstatic) [inline src/common/expression/AggregateExpression.cpp:68]: DCHECK(!!result_);

File: src/common/expression/AggregateExpression.cpp (hunk 1 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.cpp:65]: got it.

File: src/common/expression/AggregateExpression.cpp (hunk 2 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.cpp]: Fixed.

**Review by nevermore3 — DISMISSED** [review]

File: src/common/expression/AggregateExpression.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        return aggData_;
+    }
+
+    void setAggData(AggData* agg_data) {
+        aggData_ = agg_data;
+    }
+
+
+private:
+    std::function<void(AggData*, const Value&)> apply() {
```
Comment (CPWstatic) [inline src/common/expression/AggregateExpression.h:194]:
```
std::function<void(AggData*, const Value&)>   apply_;
```
No need to find the function when eval, find it at the beggining.

**Review by CPWstatic — COMMENTED** [review]
You could simply remove the AggFunction to tell the difference after the refactoring.

File: src/common/expression/AggregateExpression.h [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+    Value& deviation() {
+        return deviation_;
+    }
+
+    void setDeviation(Value deviation) {
+        deviation_ = deviation;
+    }
+
+    const Value& res() const {
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.h:77]: Just use the full word

File: src/common/expression/AggregateExpression.h [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+        return result_;
+    }
+
+    Value& res() {
+        return result_;
+    }
+
+    void setRes(Value res) {
+        result_ = res;
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.h:86]: result_ = std::move(res);

File: src/common/expression/AggregateExpression.h [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+    const Set* uniques() const {
+        return uniques_.get();
+    }
+
+    Set* uniques() {
+        return uniques_.get();
+    }
+
+    void setRes(Set* uniques) {
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.h:97]: setUniques

File: src/common/expression/test/EncodeDecodeTest.cpp [hunk 8]
```diff
@@ -146,6 +147,15 @@ TEST(ExpressionEncodeDecode, FunctionCallExpression) {
 }
 
 
+TEST(ExpressionEncodeDecode, AggregateExpression) {
```
Comment (laura-ding) [inline src/common/expression/test/EncodeDecodeTest.cpp:150]: add test for `toString()` and `clone()`

File: src/common/expression/AggregateExpression.h [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+    const Value& cnt() const {
+        return cnt_;
+    }
+
+    Value& cnt() {
+        return cnt_;
+    }
+
+    void setCnt(Value cnt) {
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.h:37]: when use?

File: src/common/expression/AggregateExpression.h (hunk 7 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:97]: Fixed, thx.

File: src/common/expression/AggregateExpression.h (hunk 5 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:77]: ok.

File: src/common/expression/AggregateExpression.h (hunk 6 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:86]: Got it.

File: src/common/expression/AggregateExpression.h (hunk 9 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:37]: Not being used.

File: src/common/expression/test/EncodeDecodeTest.cpp (hunk 8 above)
Comment (czpmango) [reply src/common/expression/test/EncodeDecodeTest.cpp:150]: Already add.

File: src/common/expression/AggregateExpression.h (hunk 4 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:198]: Got it.

File: src/common/expression/AggregateExpression.h [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    void setUniques(Set* uniques) {
+        uniques_.reset(uniques);
+    }
+
+private:
+    Value cnt_;
+    Value sum_;
+    Value avg_;
+    Value deviation_;
+    Value result_;
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.h:110]: need to init by `Value::kNullValue`

**Review by CPWstatic — APPROVED** [review]
LGTM.

**Comment by czpmango** [comment]
> You could simply remove the AggFunction to tell the difference after the refactoring.

Already update.

File: src/common/expression/AggregateExpression.h (hunk 10 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:110]: I want to initialize only with an explicit constructor.

File: src/common/expression/test/ExpressionTest.cpp [hunk 11]
```diff
@@ -201,6 +202,45 @@ class ExpressionTest : public ::testing::Test {
...
+                     const char* expr,
+                     std::vector<std::pair<std::string, Value>> inputVar,
+                     const std::unordered_map<std::string, Value> &expected) {
+        auto agg = new std::string(name);
+        auto func = new std::string(expr);
+        Expression* arg = nullptr;
+        auto isConst = false;
+        if (!func->compare("isConst")) {
+            isConst = true;
+            arg = new ConstantExpression();
```
Comment (CPWstatic) [inline src/common/expression/test/ExpressionTest.cpp:217]: delete func;

**Review by CPWstatic — DISMISSED** [review]
LGTM.

**Review by CPWstatic — APPROVED** [review]

**Review by jievince — APPROVED** [review]
LGTM

File: src/common/expression/AggregateExpression.h (hunk 4 above)
Comment (yixinglu) [reply src/common/expression/AggregateExpression.h:198]: why not change the function signature to `void apply(AggData*, const Value&)` and call the function in `apply` directly?

File: src/common/expression/AggregateExpression.h [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        kBitOr,
+        kBitXor,
+        kCollect,
+        kCollectSet,
+    };
+    static std::unordered_map<AggregateExpression::Function,
+        std::function<void(AggData*, const Value&)>> aggFunMap_;
+    static std::unordered_map<std::string, AggregateExpression::Function> nameIdMap_;
+
+    AggregateExpression(std::string* name = nullptr,
```
Comment (yixinglu) [inline src/common/expression/AggregateExpression.h:136]: explicit

File: src/common/expression/AggregateExpression.h [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        kStdev,
+        kBitAnd,
+        kBitOr,
+        kBitXor,
+        kCollect,
+        kCollectSet,
+    };
+    static std::unordered_map<AggregateExpression::Function,
+        std::function<void(AggData*, const Value&)>> aggFunMap_;
+    static std::unordered_map<std::string, AggregateExpression::Function> nameIdMap_;
```
Comment (yixinglu) [inline src/common/expression/AggregateExpression.h:134]: follow the public static variable code style: `NAME_ID_MAP` and `AGG_FUN_MAP`

File: src/common/expression/ConstantExpression.h [hunk 14]
```diff
@@ -31,6 +31,10 @@ class ConstantExpression : public Expression {
         return val_;
     }
 
+    void setValue(Value val) {
+        val_ = std::move(val);
+    }
```
Comment (yixinglu) [inline src/common/expression/ConstantExpression.h:36]: Why we need this setter function? I mean this `val_` only saves the result of evaluation and shouldn't be changed by some functions outside.

File: src/common/expression/test/ExpressionTest.cpp [hunk 15]
```diff
@@ -201,6 +202,45 @@ class ExpressionTest : public ::testing::Test {
         // EXPECT_EQ(eval.type(), expected.type());
         EXPECT_EQ(eval, expected);
     }
+
+    void testAggExpr(const char* name,
+                     const bool isDistinct,
```
Comment (yixinglu) [inline src/common/expression/test/ExpressionTest.cpp:207]: const qualifier is useless for parameter passed by value, delete it.

File: src/common/expression/test/ExpressionTest.cpp [hunk 16]
```diff
@@ -201,6 +202,45 @@ class ExpressionTest : public ::testing::Test {
...
+        auto isConst = false;
+        if (!func->compare("isConst")) {
+            isConst = true;
+            arg = new ConstantExpression();
+        } else {
+            arg = new FunctionCallExpression(func.release());
+        }
+        AggregateExpression aggExpr(agg, arg, isDistinct);
+        std::unordered_map<std::string, std::unique_ptr<AggData>> agg_data_map;
+        for (auto row : inputVar) {
```
Comment (yixinglu) [inline src/common/expression/test/ExpressionTest.cpp:223]: `const auto &row`

File: src/common/expression/test/ExpressionTest.cpp [hunk 17]
```diff
@@ -201,6 +202,45 @@ class ExpressionTest : public ::testing::Test {
...
+            auto iter = agg_data_map.find(row.first);
+            if (iter == agg_data_map.end()) {
+                agg_data_map[row.first] = std::make_unique<AggData>();
+            }
+            if (isConst) {
+                static_cast<ConstantExpression*>(arg)->setValue(row.second);
```
Comment (yixinglu) [inline src/common/expression/test/ExpressionTest.cpp:229]: I don't understand this logic, will the arg value not be reset in different loop round ?

File: src/common/expression/AggregateExpression.cpp [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+    auto val = arg_->eval(ctx);
+    auto uniques = aggData_->uniques();
+    if (distinct_) {
+        if (uniques->contains(val)) {
+            return aggData_->result();
+        }
+        uniques->values.emplace(val);
+    }
+
+    apply()(aggData_, val);
```
Comment (yixinglu) [inline src/common/expression/AggregateExpression.cpp:61]: prefer `apply(aggData_, val)`

File: src/common/expression/test/ExpressionTest.cpp (hunk 17 above)
Comment (czpmango) [reply src/common/expression/test/ExpressionTest.cpp:229]: I want to test the eval behavior when nested within AggregateExpression and not just agg_function, such as avg(abs($-.age)), where the iteration data $-.age(the input parameter to arg) is variable. Similar to emulating AggregateExecutor behavior.

File: src/common/expression/AggregateExpression.cpp (hunk 18 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.cpp:61]: Got it, thx~

File: src/common/expression/test/ExpressionTest.cpp (hunk 16 above)
Comment (czpmango) [reply src/common/expression/test/ExpressionTest.cpp:223]: ok.

File: src/common/expression/AggregateExpression.h (hunk 13 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:134]: Got it!

File: src/common/expression/ConstantExpression.h (hunk 14 above)
Comment (czpmango) [reply src/common/expression/ConstantExpression.h:36]: Similar to FunctionCallExpression SetArgs. For example, when ConstantExpression is tested as another Expression parameter, I want to just modify the input data without having to reconstruct a new ConstantExpression.

Comment (czpmango) [reply src/common/expression/ConstantExpression.h:36]: Setting the val_ of ConstantExpression dynamically can be convenient in some test scenarios.

File: src/common/expression/AggregateExpression.h (hunk 4 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h]: > why not change the function signature to `void apply(AggData*, const Value&)` and call the function in `apply` directly?
Already modified.

**Review by yixinglu — DISMISSED** [review]

**Review by jievince — DISMISSED** [review]

**Review by jievince — APPROVED** [review]

File: src/common/expression/AggregateExpression.cpp [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+                return;
+            }
+            if (UNLIKELY(val.isBadNull() || (!val.isNull() && !val.empty() && !val.isNumeric()))) {
+                res = Value::kNullBadType;
+                return;
+            }
+
+            auto& sum = aggData->sum();
+            auto& cnt = aggData->cnt();
+            if (val.isNull() || val.empty()) {
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.cpp:147]: it can move lin137

File: src/common/expression/AggregateExpression.cpp [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+            }
+            if (UNLIKELY(!val.isNull() && !val.empty() && !val.isNumeric())) {
+                res = Value::kNullBadType;
+                return;
+            }
+            if (val.isBadNull()) {
+                res = Value::kNullBadData;
+                return;
+            }
+            if (val.isNull() || val.empty()) {
```
Comment (laura-ding) [inline src/common/expression/AggregateExpression.cpp:176]: ditto

**Review by laura-ding — APPROVED** [review]

File: src/common/expression/AggregateExpression.cpp (hunk 19 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.cpp:147]: isBadNull check?

**Comment by bright-starry-sky** [comment]
This PR is related to which PR of QE? I see that the QE layer compile failed.

**Comment by czpmango** [comment]
> This PR is related to which PR of QE? I see that the QE layer compile failed.

[https://github.com/vesoft-inc/nebula-graph/pull/503](https://github.com/vesoft-inc/nebula-graph/pull/503)

**Comment by bright-starry-sky** [comment]
> > This PR is related to which PR of QE? I see that the QE layer compile failed.
> 
> [vesoft-inc/nebula-graph#503](https://github.com/vesoft-inc/nebula-graph/pull/503)

OK, thanks. It looks like it'll be ready soon.

**Comment by czpmango** [comment]
> > > This PR is related to which PR of QE? I see that the QE layer compile failed.
> > 
> > 
> > [vesoft-inc/nebula-graph#503](https://github.com/vesoft-inc/nebula-graph/pull/503)
> 
> OK, thanks. It looks like it'll be ready soon.

Already done.

**Comment by bright-starry-sky** [comment]
> > > > This PR is related to which PR of QE? I see that the QE layer compile failed.
> > > 
> > > 
> > > [vesoft-inc/nebula-graph#503](https://github.com/vesoft-inc/nebula-graph/pull/503)
> > 
> > 
> > OK, thanks. It looks like it'll be ready soon.
> 
> Already done.

👍

**Pull request merged**